### PR TITLE
Clear out ClientCache when UI is detached to prevent a minor memory leak

### PR DESCRIPTION
**Pull request by Ansku**
Implemented with a listener rather than direct call from UI.detach() in
order to avoid new public API, since the whole feature has been marked
for removal. This doesn't yet prevent the cache or the type map from
getting slightly bloated during the UI's lifetime.

See: #3705

---
This change is [Reviewable](https://reviewable.io/reviews/vaadin/framework/12199)

### CODE REVIEW

**Review by OlliTietavainenVaadin — APPROVED** [review]
:lgtm:

> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/vaadin/framework/12199#-:-MTGMva47U2kkypBFTvq:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved
